internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,311 @@
+				newAddr := resolver.Address{Addr: proxyAddr}
+				newAddr = proxyattributes.Populate(newAddr, proxyattributes.Options{
+					User:        r.proxyURL.User,
+					ConnectAddr: targetAddr.Addr,
+				})
+				addrs = append(addrs, newAddr)
+			}
+			endpoints = append(endpoints, resolver.Endpoint{Addresses: addrs})
+		}
+	}
+	return addresses, endpoints
+}
+
+// updateProxyResolverState updates the proxy resolver state by storing proxy
+// addresses and endpoints, marking the resolver as ready, and triggering a
+// state update if both proxy and target resolvers are ready. It is a
+// StateListener function of wrappingClientConn passed to the proxy resolver.
+func (r *delegatingResolver) updateProxyResolverState(state resolver.State) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if logger.V(2) {
+		logger.Infof("Addresses received from proxy resolver: %s", state.Addresses)
+	}
+	r.proxyAddrs = state.Addresses
+	r.proxyEndpoint = state.Endpoints
+	r.proxyResolverReady = true
+	r.updateStateIfReadyLocked()
+	return nil
+}
+
+// updateTargetResolverState updates the target resolver state by storing target
+// addresses, endpoints, and service config, marking the resolver as ready, and
+// triggering a state update if both resolvers are ready. It is a StateListener
+// function of wrappingClientConn passed to the target resolver.
+func (r *delegatingResolver) updateTargetResolverState(state resolver.State) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if logger.V(2) {
+		logger.Infof("Addresses received from target resolver: %v", state.Addresses)
+	}
+	r.targetAddrs = state.Addresses
+	r.targetEndpoint = state.Endpoints
+	r.targetResolverReady = true
+	// Update curState to include other state information, such as the
+	// service config, provided by the target resolver. This ensures
+	// curState contains all necessary information when passed to
+	// UpdateState. The state update is only sent after both the target and
+	// proxy resolvers have sent their updates, and curState has been
+	// updated with the combined addresses.
+	r.curState = state
+	r.updateStateIfReadyLocked()
+	return nil
+}
+
+// updateStateIfReadyLocked checks if both proxy and target resolvers are ready,
+// combines addresses, and updates the ClientConn state.
+func (r *delegatingResolver) updateStateIfReadyLocked() {
+	if !r.targetResolverReady || !r.proxyResolverReady {
+		return
+	}
+	r.curState.Addresses, r.curState.Endpoints = r.combinedAddressesLocked()
+	r.cc.UpdateState(r.curState)
+}
+
+// wrappingClientConn serves as an intermediary between the parent ClientConn
+// and the child resolvers created here. It implements the resolver.ClientConn
+// interface and is passed in that capacity to the child resolvers.
+//
+// Its primary function is to aggregate addresses returned by the child
+// resolvers before passing them to the parent ClientConn. Any errors returned
+// by the child resolvers are propagated verbatim to the parent ClientConn.
```

Comment:
This is not true anymore since the logic is now moved to the stateListener callback. So, this paragraph can be deleted.

Reply:
Done.